### ISSUE TITLE
Need Feedback for Multi Step Feature

### ISSUE THREAD

**Issue by ZafarKamal123** [munirkamal/gutenberg-forms#64]
@jwahdatehagh 
@simonhammes 
@munirkamal 
@paaljoachim

We have just added the multi step feature and merged it into the master branch at Github for now. Consider this an unfinished update as we need your **suggestions/feedback** to make it work and look perfect.

It would be great if you can test this out and provide your feedback on:

- How it Works Overall
- The UI in the editor to differentiate the steps
- Anything Else you want to comment on.

In case you wonder how it works, I have created a short rough screencast:

Link: https://share.getcloudapp.com/YEuAJjvr

Cheers 😉

**Comment by simonhammes** [comment]
Wow .. this works great already!

I have four minor suggestions:
1. When submitting the form, show the green notice above the form

2. When inserting a new Gutenberg Forms block into the editor and selecting the "Multi Step" variant, the next/submit buttons should be there by default (maybe only for the first two step blocks?)

3. I find the dropdown to alter the form button action on the right sidebar rather unobtrusive, too unobtrusive imo. Maybe we can swap the <select> out for something else    

4. Bonus: Maybe we can add more explanations when inserting the Gutenberg Forms Block about the Multi Step variant; what it is, what it allows you to do etc (maybe this is unnecessary?)

Let me know what you think :sweat_smile:

**Comment by ZafarKamal123** [comment]
Thanks for the suggestion.

1) I can add the success message above the form, But what if some user want it below the form?

2) This was in my mind too. I will try adding default buttons in the steps to make it more easy!

Let me think about the last suggestion 😉

**Comment by paaljoachim** [comment]
Wow that is cool! I just saw the video.
Here one can actually create a survey using the multistep.
I will test it out in a few minutes.

**Comment by paaljoachim** [comment]
Going through the new screens.

As one adds the block there are two options.

![Screen Shot 2020-03-23 at 19 17 41](https://user-images.githubusercontent.com/5323259/77349227-fc800300-6d3a-11ea-99c9-b64b4c8917b1.png)

Standard
Multi Step - one might not know what this means. Here it would be good with a tooltip.

Then again I am not sure if there should be an initial option like this. It is better to begin with the form directly. Perhaps even has a placeholder to where people can choose what kind of form they want. Kinda like this closed issue: https://github.com/WordPress/gutenberg/issues/20920


Choosing Standard brings up a default form.
Here it would be nice to be able to add a block field named multistep. As one could create a very long form and add the multistep block to every location in the form where one wants to split it up. 

-----

I looked for ways to go from Standard to Multistep but did not locate any.

Looking at Multistep I would say it is not needed as one could just instead add a multistep block. One adds it at the start of where one wants multistep to begin.

Kinda like:
Name
Age

- Multistep -

Occupation
How long in occupation

- multistep -

State
how long lived in state

- multistep -

Another question.
Another question.

- multistep -

That would mean having one form, and not two different separate areas.

**Comment by simonhammes** [comment]
Thank you for you reply :+1: 

> 1. I can add the success message above the form, But what if some user want it below the form?

I don't know whether we should add too many custom options for _everything_. I just thought about the way WordPress Core and Gutenberg show notices.

> 2. This was in my mind too. I will try adding default buttons in the steps to make it more easy!

Thank you :rocket: 

> Let me think about the last suggestion :wink: 

I thought about a ButtonGroup but the texts are probably too long

**Comment by munirkamal** [comment]
Hi,

So we made some tweaks to the Multi-Step as per some feedback. It is still not perfect and that is what we need to think about.

**Point One:** 

The point @paaljoachim raised about the possibility to let multiple multi-step blocks in the standard form was something we considered earlier. However, there are some issues with that.

We will have to add a Multi Steps Block which will then have a nested block Form Step (which is there now and is a single step wrapper) then users can add form field blocks inside it. 

So that makes it really complex and then we also need to think about the progress indicator which can conflict with those multiple steps blocks. 

I have also not come across any such form yet which has multiple multi-steps. @paaljoachim  Can you share any real form example which has this functionality? 

**Point Two:**

Now if we agree to how it is implemented now, there is again something to note.  Like @paaljoachim raised it, A user may insert any block between the Form Step blocks. Right now I found no easy way to limit the users to no do that. 

So, can we leave it as it is and it won't break anything but it will not work on the front-end? We can prepare proper documentation on how to create a Multi-Step Form properly so users may follow that.

Additionally, we might be able to display some warning if a user inserts any other block between the Form Step blocks to warn him that it will break the form.

What do you guys think?

**Point Three:**

The Button Block now has an "Action" setting in the block menu as well. It is to make it easier to apply certain actions from the editor directory. This is added as per @simonhammes feedback about the action control in the sidebar.

Is that good now?

**Point Four:**

The Steps separator is now tweaked as per @paaljoachim suggestions. Now it doesn't show Step end indicator.

Is that fine? Any other changes to the indicator or UI for the Steps representation in the editor? 

**Point Five:**

About the buttons. Now, this is also something users need to take care of. We keep it this way to let users structure any layout they want for each step and overall form. Unlike other multi-step forms, the next/previous and submit buttons are not limited to appear at the bottom.

Users can actually add the buttons anywhere they wish to. And then set the button text to anything they like.

Now, this also leaves users to take care of much inserting the required buttons where needed. Like, on the first step user needs to only put the Next button. On the other steps (except last) there should be Next & Previous buttons and then on the last step, there should be the Previous & Submit button.

This is all in control of a user to take care of. So what do you think about this? We will cover this in the documentation too. 

---------

That's about it. Thank you for reading this long comment till End. 😇🙏
Looking forward to your response on this to finalize this feature.

**Comment by paaljoachim** [comment]
**Point One:**
I remember some time ago testing out a multi step / multi page feature in Caldera Forms. 
I noticed one could add a button and use an option type Next Page. It then showed a page 2 to where one could add content. Basically one could just add another button and use the option type Next Page. I did a search on Youtube for Caldera Forms multi page and opened up a video there and noticed how it was done. 

https://www.youtube.com/watch?v=R_iBy8NBdB4

------------


**Point Two:**

> A user may insert any block between the Form Step blocks. Right now I found no easy way to limit the users to no do that.

Adding a block after the Next button. Then the block just shows up on the frontend after the NEXT button.  No harm done. If the user wants it that way then they leave it like that or else they will just move it upward. 

![Screen Shot 2020-03-24 at 15 17 57](https://user-images.githubusercontent.com/5323259/77435482-a6ff3100-6de2-11ea-9d82-4f9fb0c4c24c.png)

Message seen after the NEXT button.

Improvements:
- Be able to turn off all caps. So one can have "Next" 
- Option to rename NEXT/PREVIOUS/etc to something else.
- Rounding of button corners.

-------------------

**Point Three:** 
Looks good.

-----------

**Point Four:**

Step Start looks and works nicely.

There is no clear way to how to add another "Step Start"
I noticed that clicking the second inserter the "Form Step" block is seen in Gutenberg Forms area. This can be added to create another step.
I also had to add the previous and next button.
It seemed to work pretty well.
It needs testing and through it adjusting, but it is a good first version.

There are two block inserters below the Next button.
![Screen Shot 2020-03-24 at 15 24 58](https://user-images.githubusercontent.com/5323259/77436322-a0bd8480-6de3-11ea-80db-de89ca94077a.png)

Ahh hmmm....
One is for Form Step
Two is for  Gutenberg Forms
I do not know what the difference is here. I do though see the Form Step in the Gutenberg Forms Block Inserter area. Form Step is not seen in the Form Step Block inserter area.

--------

**Point Five:**

Good points. Give the user the option to where they want the buttons.

Bottom line as I see it is that this is a very nice first version. Through testing one will notice the need to adjust specific things.

**Comment by munirkamal** [comment]
@paaljoachim 

Thanks for your feedback. Just to make it clear for your point 2 feedback. You may already change the button text to whatever you like and so can adjust the case (capital, small, etc).

![Screen Recording 2020-03-25 at 04 58 PM](https://user-images.githubusercontent.com/5817882/77534240-198f1000-6eba-11ea-8056-964b1618c9f0.gif)

**Comment by munirkamal** [comment]
@paaljoachim 

Also, thank you for finding the Caldera Forms video. So that is actually working the same was as we have implemented it. It is still a single Multi-Step form. There can not be multiple multi-steps.

So I think we are good at the way we implemented it?

**Comment by paaljoachim** [comment]
Hi Munir!

Yes. I believe we are good to go!

**Comment by munirkamal** [comment]
Released 🙏
